### PR TITLE
Improving prefix-string parsing (fixes  #2339)

### LINKED ISSUE #2339: old print syntax throws
My fingers stupidly typed `print` syntax from Python 2.17. LPython responded with a barf, but, perhaps a friendly error message would be more appropriate.

repro:

```python
def main():
    print "done!"


if __name__ == '__main__':
    main()
```

result:

```
(base) ┌─(~/CLionProjects/lpython/lasr/LP-pycharm)───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(brian@MacBook-Pro:s003)─┐
└─(13:47:03 on lasr ✭)──> lpython -I. ISSUES/IssueStringMain.py                                                                                                          1 ↵ ──(Tue,Sep26)─┘
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/brian/CLionProjects/lpython/src/bin/lpython.cpp", line 1888
    err = compile_python_to_object_file(arg_file, tmp_o, runtime_library_dir,
  File "/Users/brian/CLionProjects/lpython/src/bin/lpython.cpp", line 786
    LCompilers::Result<LCompilers::LPython::AST::ast_t*> r = parse_python_file(
  File "/Users/brian/CLionProjects/lpython/src/lpython/parser/parser.cpp", line 126
    Result<LPython::AST::Module_t*> res = parse(al, input, prev_loc, diagnostics);
  File "/Users/brian/CLionProjects/lpython/src/lpython/parser/parser.cpp", line 22
    p.parse(s, prev_loc);
  File "/Users/brian/CLionProjects/lpython/src/lpython/parser/parser.cpp", line 53
    m_tokenizer.set_string(inp, prev_loc);
  File "parser.yy", line 1107
  File "/Users/brian/CLionProjects/lpython/src/lpython/parser/semantics.h", line 879
    throw LCompilers::LCompilersException("The string is not recognized by the parser.");
LCompilersException: The string is not recognized by the parser.
```

### PR DESCRIPTION
**Pull request by mrdaybird**
Fixes #2339 
Modifies how prefix-strings is parsed and tokenized. 
1. Tokenize prefixes ( 'r', 'b' etc) as Soft keyword KW_STR_PREFIX
2. Update grammar to check for KW_STR_PREFIX followed by STRING

Now the compiler gives a syntax error when encountering a identifier(not a prefix) followed by a string literal.

### CODE REVIEW

**Comment by certik** [comment]
Can you write a test that failed (or gave a misleading error message) in master, and now works (gives a good error message)?

If it is just a better error message, you can put it into `tests/error`.

**Comment by mrdaybird** [comment]
Added the tests.
But to clarify, should I use the 'ast' flag or the 'ast_new' flag for the tests? are they even relevant for testing errors?

File: tests/errors/prefix_string_01.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+def main():
+	# python2 syntax should result in a syntax error
+	print "Hello", "World"
```
Comment (certik) [inline tests/errors/prefix_string_01.py:3]: Please use 4 spaces for indentation.

File: tests/errors/prefix_string_02.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+from lpython import i32
+
+# fix difference between lpython and cpython in prefix string grammar
+# Prefix should be attached to the quote with any whitespace.
```
Comment (certik) [inline tests/errors/prefix_string_02.py:4]:
```suggestion
# Prefix should be attached to the quote without any whitespace.
```
?

File: tests/errors/prefix_string_02.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+
+def main():
+	print(r "Hello World")
```
Comment (certik) [inline tests/errors/prefix_string_02.py:7]: Use 4 spaces for indentation.

File: tests/reference/ast-prefix_string_02-3d530b2.stderr [hunk 4]
```diff
@@ -0,0 +1,5 @@
+syntax error: Token '"Hello World"' (of type 'string') is unexpected here
+ --> tests/errors/prefix_string_02.py:7:10
+  |
+7 |  print(r "Hello World")
+  |          ^^^^^^^^^^^^^
```
Comment (certik) [inline tests/reference/ast-prefix_string_02-3d530b2.stderr:5]: For comparison, this test passes in main, and in CPython 3.10.2 it gives:
```console
  File "/Users/ondrej/repos/lpython/a.py", line 7
    print(r "Hello World")
            ^^^^^^^^^^^^^
SyntaxError: invalid syntax
```
So this new LPython error is a good one.

File: tests/reference/ast-prefix_string_01-cf221fd.stderr [hunk 5]
```diff
@@ -0,0 +1,5 @@
+syntax error: Token '"Hello"' (of type 'string') is unexpected here
+ --> tests/errors/prefix_string_01.py:3:8
+  |
+3 |  print "Hello", "World" 
+  |        ^^^^^^^
```
Comment (certik) [inline tests/reference/ast-prefix_string_01-cf221fd.stderr:5]: In main we get:
```console
$ lpython b.py 
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/ondrej/repos/lpython/src/bin/lpython.cpp", line 1872
    err = compile_python_to_object_file(arg_file, tmp_o, runtime_library_dir,
  File "/Users/ondrej/repos/lpython/src/bin/lpython.cpp", line 771
    LCompilers::Result<LCompilers::LPython::AST::ast_t*> r = parse_python_file(
  File "/Users/ondrej/repos/lpython/src/lpython/parser/parser.cpp", line 126
    Result<LPython::AST::Module_t*> res = parse(al, input, prev_loc, diagnostics);
  File "/Users/ondrej/repos/lpython/src/lpython/parser/parser.cpp", line 22
    p.parse(s, prev_loc);
  File "/Users/ondrej/repos/lpython/src/lpython/parser/parser.cpp", line 53
    m_tokenizer.set_string(inp, prev_loc);
  File "parser.yy", line 1107
  File "/Users/ondrej/repos/lpython/src/lpython/parser/semantics.h", line 879
    throw LCompilers::LCompilersException("The string is not recognized by the parser.");
LCompilersException: The string is not recognized by the parser.
```
and in CPython 3.10.2 we get:
```console
  File "/Users/ondrej/repos/lpython/b.py", line 3
    print "Hello", "World" 
    ^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)?
```
So the above error is a good one. 

We could later on improve the parser to even parse the old syntax, but it's probably not worth it, as long as it gives a nice error message now.

**Comment by certik** [comment]
This looks great, thanks for fixing it and adding tests. I left some minor comments to fix.


>  But to clarify, should I use the 'ast' flag or the 'ast_new' flag for the tests? are they even relevant for testing errors?

Use `ast` to test syntax errors. Use `asr` to test semantic errors.

**Comment by certik** [comment]
You have to update reference tests via `./run_tests -u`.

**Comment by mrdaybird** [comment]
Thanks I noticed a little late. :sweat_smile: 
Should I rebase as well?

**Comment by certik** [comment]
Either polish the commits to a few logical commits and rebase, or just leave it and I'll squash and merge it.

**Comment by mrdaybird** [comment]
I think you can just squash and merge it. Think I would probably mess things up with the rebase.

**Review by certik — APPROVED** [review]

**Pull request merged**